Feature/opendistro security plugin

SolomonShorser-OICR opened the pull request:
Enables security with the OpenDistro for Elasticsearch security plugin (https://opendistro.github.io/for-elasticsearch-docs/docs/security/).

SolomonShorser-OICR committed:
Successfully got OpenDistro-for-Elasticsearch working

SolomonShorser-OICR committed:
Include other key files in Kibana; stop using roles mapping

SolomonShorser-OICR committed:
Version upgrade

SolomonShorser-OICR committed:
Update opendistro_security.authcz.admin_dn

...as per the advice from the securityadmin script

SolomonShorser-OICR committed:
Enable monitoring; include SSL cert/key

SolomonShorser-OICR committed:
Fixed spelling error

SolomonShorser-OICR committed:
logstash needs root ca

SolomonShorser-OICR committed:
better self-signed certs

SolomonShorser-OICR committed:
Add logstash user.

SolomonShorser-OICR committed:
Tweak some settings (not sure it helped much).

SolomonShorser-OICR committed:
Enable SSL and user authentication for logstash pipeline

SolomonShorser-OICR committed:
Sync version with other components.

SolomonShorser-OICR committed:
Add note RE: logstash user

SolomonShorser-OICR committed:
New roles.yml file.

SolomonShorser-OICR committed:
include specific keys instead of blocking.

Blocking extracted keys was getting too complicated and the list was
getting too big, so now instead, use a list of allowed keys.

SolomonShorser-OICR committed:
nginx is no longer needed

SolomonShorser-OICR committed:
cleanup

SolomonShorser-OICR committed:
Update README, add utility script.

SolomonShorser-OICR committed:
cleanup

SolomonShorser-OICR committed:
Add documentation

SolomonShorser-OICR committed:
Update documentation

SolomonShorser-OICR committed:
Added section on  setting password.

SolomonShorser-OICR committed:
update numbers

SolomonShorser-OICR committed:
Whitelist didn't quite work...

Whitelist needs to be able to include everything that is not a
"kvXtrct_" field.

SolomonShorser-OICR committed:
Add note about certificate expiry

cookersjs commented on docker-compose.yaml:
```diff
@@ -1,28 +1,67 @@
 version: '3'
 
+# Generate self-signed certs with:
+# openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 365 -nodes
+# openssl req -new -key key.pem -x509 -days 1000 -out CA-cert.pem
```
What is the different between these 2 certificates?

SolomonShorser-OICR replied:
A CA certificate is a certificate from a "Certificate Authority". It can be used to generate other certificates. Of course, since this is self-signed and generated, most web browsers would throw up a lot of red flags. If the services using these certificates were to be opened up the rest of the world, I'd probably be using certificates from Let's Encrypt, or something like that. But since this is all internal, self-signed certificates are OK. They can still be used to encrypt traffic, they're just not as secure-looking as certificates from an actual CA.

cookersjs commented on docker-compose.yaml:
```diff
@@ -1,28 +1,67 @@
 version: '3'
 
+# Generate self-signed certs with:
+# openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -days 365 -nodes
+# openssl req -new -key key.pem -x509 -days 1000 -out CA-cert.pem
+
+# # Generating certificates: https://opendistro.github.io/for-elasticsearch-docs/docs/security-configuration/generate-certificates/
+# # First two commands result in root CA file: root-ca.pem
+# openssl genrsa -out root-ca-key.pem 2048
+# openssl req -new -x509 -sha256 -key root-ca-key.pem -out root-ca.pem
+# # Next few commands generate key (admin-key.pem) and certificate (admin.pem)
+# openssl genrsa -out admin-key-temp.pem 2048
+# openssl pkcs8 -inform PEM -outform PEM -in admin-key-temp.pem -topk8 -nocrypt -v1 PBE-SHA1-3DES -out admin-key.pem
+# openssl req -new -key admin-key.pem -out admin.csr
+# openssl x509 -req -in admin.csr -CA root-ca.pem -CAkey root-ca-key.pem -CAcreateserial -sha256 -out admin.pem
+# #...
+# #Then export JAVA_HOME=/usr/share/elasticsearch/jdk
```
Is the whole point of this commented out section that contains commands to show what is happening with the docker-compose command, or is it to show how someone might do it manually?

SolomonShorser-OICR replied:
That is a note about setting  up the certificates manually. I've since added that to the README.md so this can probably be removed from the docker-compose file.

jweiser replied:
I noticed the same as @cookersjs - has or can this section of comments be removed?

SolomonShorser-OICR replied:
This is cleaned up now.

cookersjs commented on docker-compose.yaml:
```diff
@@ -1,28 +1,67 @@
+            - "./root-ca.pem:/usr/share/elasticsearch/config/root-ca.pem:ro"
+            - "./internal_users.yml:/usr/share/elasticsearch/plugins/opendistro_security/securityconfig/internal_users.yml"
+            - "./roles.yml:/usr/share/elasticsearch/plugins/opendistro_security/securityconfig/roles.yml"
+            # - "./roles_mapping.yml:/usr/share/elasticsearch/plugins/opendistro_security/securityconfig/roles_mapping.yml"
```
Should this be removed?

SolomonShorser-OICR replied:
Not sure. Custom roles mappings might be wanted/needed in the future. Currently, the role for Logstash needs to be manually tweaked in the UI. Theoretically, it should be possible to do it in the roles-mapping.yaml file, but I haven't had a chance to work on that in detail yet. And since it's not that difficult to do in the UI, I didn't make that a priority.

cookersjs commented on elasticsearch.yml:
```diff
@@ -0,0 +1,33 @@
+network.host: 0.0.0.0
+
+######## Start OpenDistro for Elasticsearch Security Configuration ########
+# WARNING: revise all the lines below before you go into production
```
What will happen if someone runs this without adjusting the lines? Would it fail or create an incorrectly configured instance?

SolomonShorser-OICR replied:
That comment is from the elasticsearch.yml config copied from the OpenDistro documentation and examples. The default file references some placeholder certificates which should not be used in production systems. Some other fields also might have default values not valid for other people's setups, so it could fail to run properly if you don't double-check the config.

cookersjs reviewed (commented):
Just have a few questions that I'd like answered. I definitely don't know enough about this to provide any kind of critical review though!

SolomonShorser-OICR committed:
Updates to reporting scripts & queries

Report script has been updated to take a username/password for elastic
from a .netrc file. This is because the OpenDistro for Elastic security
plugin requires authentication. Also updated some queries for current
version of elasticsearch.

SolomonShorser-OICR commented:
@cookersjs @jweiser Any further comments? I'd like to merge this eventually.

SolomonShorser-OICR committed:
update gitignore

jweiser commented on kibana_saved_objects/kibana_visualizations.ndjson:
```diff
@@ -1,154 +1,154 @@
-{"attributes":{"description":"","kibanaSavedObjectMeta":{"searchSourceJSON":"{\"filter\":[],\"query\":{\"query\":\"probably_a_bot:false && (reactome_download_file_name:/.*Reactome((To)|2).*/ || reactome_download_file_name:*BioPax* || reactome_download_file_name:*ReactomePathways.gmt*) && (response:2*)\",\"language\":\"lucene\"},\"indexRefName\":\"kibanaSavedObjectMeta.searchSourceJSON.index\"}"}, ...
-{"attributes":{"description":"","kibanaSavedObjectMeta":{"searchSourceJSON":"{\"filter\":[],\"query\":{\"query\":\"_exists_:reactome_content_service_exporter_path  && !response:5*\",\"language\":\"lucene\"},\"indexRefName\":\"kibanaSavedObjectMeta.searchSourceJSON.index\"}"},"title":"ContentService exporters - bar graph","uiStateJSON":"{}","version":1,"visState":"{\"title\":\"ContentService expor ...
-{"attributes":{"description":"","kibanaSavedObjectMeta":{"searchSourceJSON":"{\"query\":{\"query\":\"probably_a_bot:false && (response:200 || response:304) && internal_IP:false && !(clientip:127.0.0.1 || clientip:\\\"::1\\\" || clientip:206.108.125.163 || clientip:100.25.71.177) && programmatic_access:false && !request:/.*\\\\\\\\.((css)|(js)|(class)|(gif)|(jpg)|(jpeg)|(png)|(bmp)|(ico)|(rss)|(xm ...
+{"attributes":{"description":"","kibanaSavedObjectMeta":{"searchSourceJSON":"{\"query\":{\"query\":\"(response:4* || response:5* ) && referrer_domain:*.best\",\"language\":\"lucene\"},\"filter\":[],\"indexRefName\":\"kibanaSavedObjectMeta.searchSourceJSON.index\"}"},"title":"Error Analysis - counts of requests per .best domain","uiStateJSON":"{\"vis\":{\"params\":{\"sort\":{\"columnIndex\":null,\ ...
```
Is this .ndjson file autogenerated?

SolomonShorser-OICR replied:
This is exported from the Kibana UI.

jweiser commented on logstash.dockerfile:
```diff
@@ -1,4 +1,6 @@
-FROM docker.elastic.co/logstash/logstash:7.6.1
+# Before building this dockerfile, remember to build the logstash-filter-usage_type plugin: `cd logstash-plugins/usagetype-plugin && ./gradlew gem && mv *.gem ../../ && mv *.gemspec ../../`
+# For more on building logstash plugins, see: https://www.elastic.co/guide/en/logstash/7.x/java-filter-plugin.html#_running_the_gradle_packaging_task_3
+FROM docker.elastic.co/logstash/logstash:7.7.0
```
Is there a way in a dockerfile to include the build requirements mentioned in the comments?  

If not, maybe including a small wrapper shell script that builds the requirements and then builds the dockerfile?

SolomonShorser-OICR replied:
The  Usage Type plugin is meant to be built outside the container, for ease of development and testing. A wrapper script could be added at a future date but will not be included in this PR.

jweiser replied:
Ah okay - is a note about this requirement in or needed in the README?  I'm just thinking if someone tries to run this dockerfile before the Usage Type plugin requirement and it doesn't work but they don't think to look in the dockerfile itself for this documentation.

SolomonShorser-OICR replied:
I can add a note to the README.

SolomonShorser-OICR replied:
README is updated.

SolomonShorser-OICR committed:
Clean up docker-compose file.

jweiser commented on roles.yml:
```diff
@@ -0,0 +1,31 @@
+---
+_meta:
+  type: "roles"
+  config_version: 2
```
Does 2 mean the version of elasticsearch in this context or something else?

SolomonShorser-OICR replied:
It's for the OpenDistro For ElasticSearch plugin.

jweiser replied:
Got it - do you think it's worth adding a comment or will be clear to a user from context?

SolomonShorser-OICR replied:
See: https://opendistro.github.io/for-elasticsearch-docs/docs/security/configuration/yaml/

They do not provide a detailed explanation for "config_version", so my guess is that it tells the plugins what version of the configuration it is using.

jweiser replied:
Okay, I just meant something simple like:
```
# OpenDistro Config Version For ElasticSearch Plugin
config_version: 2 
```
Unless that's already known to a user of this code.  I'm not sure if it is or isn't.

SolomonShorser-OICR replied:
It's in all of their examples with zero explanation, so I don't want to _add_ my own explanation, in case I don't quite get it right.

jweiser replied:
Fair enough!

jweiser commented on security-config.yml:
```diff
@@ -0,0 +1,14 @@
+    authc:
+        http_enabled: true
+        transport_enabled: true
+        order: 1
```
What does "1" mean here?

SolomonShorser-OICR replied:
From the documentation: https://opendistro.github.io/for-elasticsearch-docs/docs/security/configuration/configuration/

> The order makes it possible to chain authentication domains together. The security plugin uses them in the order that you provide.

Even though we have 1 authentication domain, I think we we still need to provide a value for this.

jweiser replied:
Maybe a comment to this effect in the yml file would be good so the meaning is known?

SolomonShorser-OICR replied:
I've added a note to the file.

jweiser commented on roles_mapping.yml:
```diff
@@ -0,0 +1,73 @@
+
+_meta:
+  type: "rolesmapping"
+  config_version: 2
```
Same question as https://github.com/reactome/elk-setup/pull/9/files#r497030033

SolomonShorser-OICR committed:
Update README

SolomonShorser-OICR committed:
Added note

jweiser reviewed (approved):
Thanks for looking over my comments - looks good to me!
